src/math/math_base.h

```diff
@@ -105,5 +105,95 @@ struct PointBase3D : public PointBase2D<_Type>
     _Type z;
 };
 
+template <typename _Type>
+struct LineBase2D
+{
+    LineBase2D( const PointBase2D<_Type> & _p1, const PointBase2D<_Type> & _p2 )
+        : p1( _p1 )
+        , p2( _p2 )
+    { }
+
+    LineBase2D( const PointBase2D<_Type> & _p, _Type angleRadians )
+        : p1( _p )
+        , p2( _p.x + std::cos(angleRadians), _p.y + std::sin(angleRadians) )
+    { }
+
+    bool operator == ( const LineBase2D & line ) const
+    {
+        return parallel( line ) && pvmath::isEqual<_Type>(distance(line.p1), 0);
+    }
+
+    LineBase2D operator + ( const PointBase2D<_Type> & point )
+    {
+        return LineBase2D(p1 + point, p2 + point);
+    }
+
+    // This is translation (shift) function
+    LineBase2D & operator += ( const PointBase2D<_Type> & point )
+    {
+        p1 += point;
+        p2 += point;
+        return *this;
+    }
+
+    double angle() const
+    {
+        return std::atan2(p2.y - p1.y, p2.x - p1.x);
+    }
+
+    bool intersect( const LineBase2D & line ) const
+    {
+        PointBase2D<_Type> point;
+        return intersect( line, point );
+    }
+
+    bool intersect( const LineBase2D & line, PointBase2D<_Type> & point ) const
+    {
+        // based on Graphics Gems III, Faster Line Segment Intersection, p. 199-202
+        // http://www.realtimerendering.com/resources/GraphicsGems/gems.html#gemsiii
+        const PointBase2D<_Type> a = p2 - p1;
+        const PointBase2D<_Type> b = line.p1 - line.p2;
+        const PointBase2D<_Type> c = p1 - line.p1;
+
+        const double denominator = a.y * b.x - a.x * b.y;
+        if (pvmath::isEqual<_Type>(denominator, 0))
+            return false; // parallel
+
+        const double na = (b.y * c.x - b.x * c.y) / denominator;
+        point = p1 + a * na;
+        return true;
+    }
+
+    bool parallel( const LineBase2D & line )
+    {
+        const PointBase2D<_Type> A = p2 - p1;
+        const PointBase2D<_Type> B = line.p1 - line.p2;
+        const PointBase2D<_Type> C = p1 - line.p1;
+
+        const double denominator = A.y * B.x - A.x * B.y;
+        return pvmath::isEqual<_Type>(denominator, 0);
+    }
+
+    _Type distance( const PointBase2D<_Type> & point ) const
+    {
+        // Line equation in the Cartesian coordinate system is
+        // y = a * x + b or A * x + B * y + C = 0
+        // A distance from a point to a line can be calculated as:
+        // |A * x0 + B * y0 + C| / sqrt(A * A + B * B)
+        if ( p1 == p2 ) // there is no line
+            return 0;
+        const _Type A = (p2.y - p1.y);
+        const _Type B = -(p2.x - p1.x);
+        const _Type C = (p2.x * p1.y - p1.x * p2.y);
+        
+        const _Type numerator = (A * point.x + B * point.y + C);
+        return (numerator < 0 ? -numerator : numerator) / sqrt( A * A + B * B );
```

Comment:
I personally faced a situation in the past when I had to measure a distance from many points to a line. In such situation **sqrt** makes a significant impact into performance. My suggestion is to make p1 and p2 as a private members of the class. More precisely call them as **position** and **direction**. Position remains as it is - p1, while direction is normalised difference between p2 and p1. Now what we would have: sqrt equation would be equal to 1 because it's normalized. Coefficients A and B in distance function are already calculated. Variables A and B in parallel and intersect functions are calculated as well, they are equal to direction. Profit! :)

Reply:
If you agree with my suggestion I can make changes myself to save your time